File: system-test/dotnet/Dsl.Core/UseCase/UseCaseDsl.cs
using Dsl.Port;
using Dsl.Core.External.Clock;
using Driver.Port.External.Clock;
using Dsl.Core.External.Erp;
using Driver.Port.External.Erp;
using Driver.Adapter.Shop.Api;
using Driver.Adapter.Shop.Ui;
using Dsl.Core.Shop;
using Driver.Port.Shop;
using Driver.Port.External.Tax;
using Dsl.Core.External.Tax;
using Driver.Adapter.External.Erp;
using Optivem.Testing;
using Dsl.Core.Shared;
using Driver.Adapter.External.Clock;
using Driver.Adapter.External.Tax;
using Optivem.Shop.SystemTest.Channel;

namespace Dsl.Core;

public class UseCaseDsl : IAsyncDisposable
{
    private const string StaticChannel = ChannelType.API;

    private readonly UseCaseContext _context;
    private readonly Configuration _configuration;
    private readonly Dictionary<string, ShopDsl> _shops = new();
    private ErpDsl? _erp;
    private TaxDsl? _tax;
    private ClockDsl? _clock;

    public UseCaseDsl(Configuration configuration)
    {
        _context = new UseCaseContext(configuration.ExternalSystemMode);
        _configuration = configuration;
    }

    public async Task<ShopDsl> Shop(ChannelMode mode, Channel channel)
    {
        var channelType = ResolveShopChannel(mode, channel);
        return await GetOrCreateShop(channelType);
    }

    public async Task<ShopDsl> Shop(Channel channel)
    {
        return await Shop(_configuration.ChannelMode, channel);
    }

    private async Task<ShopDsl> GetOrCreateShop(string channelType)
    {
        if (!_shops.TryGetValue(channelType, out var shop))
        {
            shop = await ShopDsl.CreateAsync(await CreateShopDriverForChannelAsync(channelType), _context);
            _shops[channelType] = shop;
        }
        return shop;
    }

    private static string ResolveShopChannel(ChannelMode mode, Channel channel)
    {
        var channelType = mode switch
        {
            ChannelMode.Static => StaticChannel,
            ChannelMode.Dynamic => channel.Type,
            _ => throw new InvalidOperationException($"Unknown channel mode: {mode}")
        };
        Console.WriteLine($"[ChannelMode] mode={mode} → channel={channelType}");
        return channelType;
    }

    public async Task<ShopDsl> ApiShop() => await GetOrCreateShop(StaticChannel);

    public ErpDsl Erp() => GetOrCreate(ref _erp, () => new ErpDsl(CreateErpDriver(), _context));

    public TaxDsl Tax() => GetOrCreate(ref _tax, () => new TaxDsl(CreateTaxDriver(), _context));

    public ClockDsl Clock() => GetOrCreate(ref _clock, () => new ClockDsl(CreateClockDriver(), _context));

    private async Task<IShopDriver> CreateShopDriverForChannelAsync(string channelType)
    {
        return channelType switch
        {
            ChannelType.UI => await ShopUiDriver.CreateAsync(_configuration.ShopUiBaseUrl),
            ChannelType.API => new ShopApiDriver(_configuration.ShopApiBaseUrl),
            _ => throw new InvalidOperationException($"Unknown channel type: {channelType}")
        };
    }

    private IErpDriver CreateErpDriver()
    {
        return _context.ExternalSystemMode switch
        {
            ExternalSystemMode.Real => new ErpRealDriver(_configuration.ErpBaseUrl),
            ExternalSystemMode.Stub => new ErpStubDriver(_configuration.ErpBaseUrl),
            _ => throw new InvalidOperationException($"Unknown external system mode: {_context.ExternalSystemMode}")
        };
    }

    private ITaxDriver CreateTaxDriver()
    {
        return _context.ExternalSystemMode switch
        {
            ExternalSystemMode.Real => new TaxRealDriver(_configuration.TaxBaseUrl),
            ExternalSystemMode.Stub => new TaxStubDriver(_configuration.TaxBaseUrl),
            _ => throw new InvalidOperationException($"Unknown external system mode: {_context.ExternalSystemMode}")
        };
    }

    private IClockDriver CreateClockDriver()
    {
        return _context.ExternalSystemMode switch
        {
            ExternalSystemMode.Real => new ClockRealDriver(),
            ExternalSystemMode.Stub => new ClockStubDriver(_configuration.ClockBaseUrl),
            _ => throw new InvalidOperationException($"Unknown external system mode: {_context.ExternalSystemMode}")
        };
    }

    public async ValueTask DisposeAsync()
    {
        foreach (var shop in _shops.Values)
            await shop.DisposeAsync();

        if (_erp != null)
            await _erp.DisposeAsync();

        _tax?.Dispose();

        if (_clock != null)
            await _clock.DisposeAsync();

        ChannelContext.Clear();
    }

    private static T GetOrCreate<T>(ref T? instance, Func<T> supplier) where T : class
    {
        return instance ??= supplier();
    }
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/UseCaseDsl.java
package com.optivem.shop.dsl.core.usecase;

import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.core.shared.UseCaseContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import com.optivem.shop.dsl.core.usecase.external.clock.ClockDsl;
import com.optivem.shop.dsl.core.usecase.external.erp.ErpDsl;
import com.optivem.shop.dsl.core.usecase.external.tax.TaxDsl;
import com.optivem.shop.dsl.core.usecase.shop.ShopDsl;
import com.optivem.shop.dsl.channel.ChannelType;
import com.optivem.shop.dsl.driver.port.external.clock.ClockDriver;
import com.optivem.shop.dsl.driver.port.external.erp.ErpDriver;
import com.optivem.shop.dsl.driver.port.external.tax.TaxDriver;
import com.optivem.shop.dsl.driver.port.shop.ShopDriver;
import com.optivem.shop.dsl.port.ChannelMode;
import com.optivem.shop.dsl.port.ExternalSystemMode;
import com.optivem.testing.contexts.ChannelContext;

import java.io.Closeable;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

public class UseCaseDsl implements Closeable {
    private static final Logger log = LoggerFactory.getLogger(UseCaseDsl.class);
    private static final String STATIC_CHANNEL = ChannelType.API;

    private final UseCaseContext context;
    private final ChannelMode channelMode;
    private final Function<String, ShopDriver> shopDriverFactory;
    private final Supplier<ErpDriver> erpDriverSupplier;
    private final Supplier<ClockDriver> clockDriverSupplier;
    private final Supplier<TaxDriver> taxDriverSupplier;

    private final Map<String, ShopDsl> shops = new HashMap<>();
    private ErpDriver erpDriver;
    private ClockDriver clockDriver;
    private TaxDriver taxDriver;

    private ErpDsl erp;
    private ClockDsl clock;
    private TaxDsl tax;

    public UseCaseDsl(ExternalSystemMode externalSystemMode,
                     ChannelMode channelMode,
                     Function<String, ShopDriver> shopDriverFactory,
                     Supplier<ErpDriver> erpDriverSupplier,
                     Supplier<ClockDriver> clockDriverSupplier,
                     Supplier<TaxDriver> taxDriverSupplier) {
        this.context = new UseCaseContext(externalSystemMode);
        this.channelMode = channelMode;
        this.shopDriverFactory = shopDriverFactory;
        this.erpDriverSupplier = erpDriverSupplier;
        this.clockDriverSupplier = clockDriverSupplier;
        this.taxDriverSupplier = taxDriverSupplier;
    }

    @Override
    public void close() {
        shops.values().forEach(Closer::close);

        if (erp != null) {
            Closer.close(erp);
        } else {
            Closer.close(erpDriver);
        }

        if (clock != null) {
            Closer.close(clock);
        } else {
            Closer.close(clockDriver);
        }

        if (tax != null) {
            Closer.close(tax);
        } else {
            Closer.close(taxDriver);
        }
    }

    public ShopDsl shop(ChannelMode mode) {
        var channel = resolveShopChannel(mode);
        return getOrCreateShop(channel);
    }

    public ShopDsl shop() {
        return shop(channelMode);
    }

    private ShopDsl getOrCreateShop(String channel) {
        return shops.computeIfAbsent(channel, ch ->
                new ShopDsl(shopDriverFactory.apply(ch), context));
    }

    private String resolveShopChannel(ChannelMode mode) {
        String channel;
        if (mode == ChannelMode.STATIC) {
            channel = STATIC_CHANNEL;
        } else if (mode == ChannelMode.DYNAMIC) {
            channel = ChannelContext.get();
        } else {
            throw new IllegalStateException("Unknown channel mode: " + mode);
        }
        log.info("[ChannelMode] mode={} → channel={}", mode, channel);
        return channel;
    }

    public ErpDsl erp() {
        return getOrCreate(erp, () -> {
            erpDriver = getOrCreate(erpDriver, erpDriverSupplier);
            erp = new ErpDsl(erpDriver, context);
            return erp;
        });
    }

    public ClockDsl clock() {
        return getOrCreate(clock, () -> {
            clockDriver = getOrCreate(clockDriver, clockDriverSupplier);
            clock = new ClockDsl(clockDriver, context);
            return clock;
        });
    }

    public TaxDsl tax() {
        return getOrCreate(tax, () -> {
            taxDriver = getOrCreate(taxDriver, taxDriverSupplier);
            tax = new TaxDsl(taxDriver, context);
            return tax;
        });
    }

    private static <T> T getOrCreate(T instance, Supplier<T> supplier) {
        return instance != null ? instance : supplier.get();
    }
}
